app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js

```diff
@@ -0,0 +1,90 @@
+        json[i].w == undefined ? undefined : allW.push(json[i].w);
+        json[i].h == undefined ? undefined : allH.push(json[i].h);
+    }
+    let totalWidth = (allX.length != json.length | allW.length != json.length) ? json.length : Math.max(...allX.map(function(num, idx) {return num+allW[idx]}));
+    let totalHeight = (allY.length != json.length | allH.length != json.length) ? 1 : Math.max(...allY.map(function(num, idx) {return num+allH[idx]}));
+
+    return [totalHeight, totalWidth]
+}
+
+function createViewportDiv(divId, viewportNum, viewportSpec, totalHeight, totalWidth) {
+    let viewportDiv = document.createElement("div");
+    viewportDiv.setAttribute('id', viewportSpec.viewport_name);
+    viewportDiv.style.opacity = '0.5';
+    viewportDiv.style.position = 'absolute';
+    viewportDiv.style.zIndex = '-' + viewportSpec.order;
+    viewportDiv.style.width = isNaN(viewportSpec.w) ? (1 / parseFloat(totalWidth).toFixed(2))*100 + '%' : (viewportSpec.w / parseFloat(totalWidth).toFixed(2))*100 + '%';
+    viewportDiv.style.height = isNaN(viewportSpec.h) ? "100%" : (viewportSpec.h / parseFloat(totalHeight).toFixed(2))*100 + '%';
+    viewportDiv.style.left = isNaN(viewportSpec.x) ? (viewportNum / parseFloat(totalWidth).toFixed(2))*100 + '%' : (viewportSpec.x / parseFloat(totalWidth).toFixed(2))*100 + '%';
+    viewportDiv.style.top = isNaN(viewportSpec.y) ? "0%" : (viewportSpec.y / parseFloat(totalHeight).toFixed(2))*100 + '%';
+    document.getElementById(divId).appendChild(viewportDiv).classList.add('bg-dark', 'rounded', 'border', 'border-2', 'd-flex', 'flex-column', 'justify-content-center', 'align-items-center');
+
+    if (/\d/.test(divId)) {
+        viewportDiv.innerHTML += '<p class="mb-0" style="font-size: 0.9em">'+ viewportSpec.viewport_name + '</p>';
+    } else {
+        viewportDiv.innerHTML += '<p style="font-size: 1.5em">' + viewportSpec.viewport_name + '</p>';
+        if (viewportSpec.fullsizable == true) {
+            viewportDiv.innerHTML += '<i class="fas fa-expand m-1"></i>'
+        }
+    }
+}
+
+function updateHangingProtocolVisualization(parentDivId, jsonString){
+    jsonString = jsonString || document.getElementById("id_json").value;
+    parentDivId = parentDivId || "hpVisualization";
+    showOrHideVisualizationDiv(parentDivId, jsonString);
+    try {
+        let jsonSpec = JSON.parse(jsonString);
+        [totalHeight, totalWidth] = getGridDimensions(jsonSpec);
+        removeAllChildNodes(document.getElementById(parentDivId));
+        for (let i = 0; i < jsonSpec.length; i++) {
+            createViewportDiv(parentDivId, i, jsonSpec[i], totalHeight, totalWidth);
+        }
+    } catch (err) {
+    }
+}
+
+$(document).ready(function () {
+    if (window.location.href.includes("update") | window.location.href.includes("create")) {
+        let jsonString = document.getElementById("id_json").value;
+        showOrHideVisualizationDiv("hpVisualization", jsonString);

```

Comment:
This line is not necessary since it is already called in the `updateHangingProtocolVisualization` function below, right?